--- a/src/main.cpp
+++ b/src/main.cpp
@@ -213,13 +213,3 @@
     reapZombieChildrenAutomatically();
-
-     bool watchdogOk = watchdogFd > 0;
-
-     if (watchdogFd > 0)
-         watchdogOk = g_pCompositor->setWatchdogFd(watchdogFd);
-     if (safeMode)
-         g_pCompositor->m_safeMode = true;
-
-     if (!watchdogOk)
-         Log::logger->log(Log::WARN, "WARNING: Hyprland is being launched without start-hyprland. This is highly advised against.");
-
     g_pCompositor->initServer(socketName, socketFd);

